File: leetcode/C++/jump-game-vi.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        int sol=0;
        deque<pair<int,int>> dq;
        for(int i=0,v=0; i<nums.size(); ++i){
            if(!dq.empty() && dq.front().first < i-k) dq.pop_front();
            v = (!dq.empty())? dq.front().second + nums[i] : nums[i];
            while(!dq.empty() && dq.back().second < v) dq.pop_back();
            dq.push_back({i,v});
            sol=v;
        }
        return sol;
    }
};

File: leetcode/Python/jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        sol, dq = 0, collections.deque()
        for i in range(len(nums)):
            if dq and dq[0][0] < i-k:
                dq.popleft()
            sol = nums[i] if not dq else nums[i]+dq[0][1]
            while dq and dq[-1][1] < sol:
                dq.pop()
            dq.append((i,sol))
        return sol